rasa/core/training/story_conflict.py

```diff
@@ -0,0 +1,292 @@
+        return hash(str(list(self.sliced_states)))
+
+
+def find_story_conflicts(
+    trackers: List[TrackerWithCachedStates], domain: Domain, max_history: int
+) -> List[StoryConflict]:
+    """Generates a list of `StoryConflict` objects, describing conflicts in the given trackers.
+
+    Args:
+        trackers: Trackers in which to search for conflicts.
+        domain: The domain.
+        max_history: The maximum history length to be taken into account.
+    Returns:
+        List of conflicts.
+    """
+    # We do this in two steps, to reduce memory consumption:
+
+    # Create a 'state -> list of actions' dict, where the state is
+    # represented by its hash
+    conflicting_state_action_mapping = _find_conflicting_states(
+        trackers, domain, max_history
+    )
+
+    # Iterate once more over all states and note the (unhashed) state,
+    # for which a conflict occurs
+    conflicts = _build_conflicts_from_states(
+        trackers, domain, max_history, conflicting_state_action_mapping
+    )
+
+    return conflicts
+
+
+def _find_conflicting_states(
+    trackers: List[TrackerWithCachedStates], domain: Domain, max_history: int
+) -> Dict[int, Optional[List[Text]]]:
+    """Identifies all states from which different actions follow.
+
+    Args:
+        trackers: Trackers that contain the states.
+        domain: The domain object.
+        max_history: Number of turns to take into account for the state descriptions.
+
+    Returns:
+        A dictionary mapping state-hashes to a list of actions that follow from each state.
+    """
+    # Create a 'state -> list of actions' dict, where the state is
+    # represented by its hash
+    state_action_mapping = defaultdict(list)
+    for element in _sliced_states_iterator(trackers, domain, max_history):
+        hashed_state = element.sliced_states_hash
+        if element.event.as_story_string() not in state_action_mapping[hashed_state]:
+            state_action_mapping[hashed_state] += [element.event.as_story_string()]
+
+    # Keep only conflicting `state_action_mapping`s
+    return {
+        state_hash: actions
+        for (state_hash, actions) in state_action_mapping.items()
+        if len(actions) > 1
+    }
+
+
+def _build_conflicts_from_states(
+    trackers: List[TrackerWithCachedStates],
+    domain: Domain,
+    max_history: int,
+    conflicting_state_action_mapping: Dict[int, Optional[List[Text]]],
+) -> List["StoryConflict"]:
+    """Builds a list of `StoryConflict` objects for each given conflict.
+
+    Args:
+        trackers: Trackers that contain the states.
+        domain: The domain object.
+        max_history: Number of turns to take into account for the state descriptions.
+        conflicting_state_action_mapping: A dictionary mapping state-hashes to a list of actions
+                                          that follow from each state.
+
+    Returns:
+        A list of `StoryConflict` objects that describe inconsistencies in the story
```

Comment:
optional: don't repeat type
```suggestion
        StoryConflict` objects that describe inconsistencies in the story
```

Reply:
Here I don't mean the type, but just that it returns 0, 1, or several of these conflicts. I.e. "list" in the generic sense, not in the Python type sense. Is there a better word?